### PR TITLE
ModeSpec, replacing invididual specs for each mode

### LINKED ISSUE #215: Refactor the `FooSpec`s to be mutually exclusive
> @bzinberg good question - I think for the initial release we'll just make `AutoDetect` monolithic, but we should probably leave open the later possibility of choosing a subset of modes. So really it would be something like
> 
> ```rust
> #[non_exhaustive]
> pub struct AutoDetectConfig {
>   // empty for now, could add power user features later
> }
> 
> pub enum ModeSpec {
>   ProvidedClassic,
>   ...,
>   AutoDetect(AutoDetectConfig),
> }
> ```
> 
> That said, I think structuring this as something like `vec![bid_a, bid_b].iter().max_by_key(|bid| bid.bits_saved)` makes sense in any case.
> 
> @Skielex Yep, the modes are exclusive. I see what you're getting at, but I think the complexity of making `ChunkConfig` generic would require a lot more code and outweight the benefits. For similar reasons, I've made e.g. `FloatMultSpec::Provided(f64)` instead of `...Provided(T)`. On incorrect use, users will get `InvalidArgument` errors, which are obviously not as good as compile-time errors, but these will still fail fast and be easy to test for.
> 
_Originally posted by @mwlon in https://github.com/mwlon/pcodec/issues/212#issuecomment-2183543586_

### PR DESCRIPTION
**Pull request by mwlon**
closes #215

### CODE REVIEW

**Comment by mwlon** [comment]
FYI @bzinberg. Very satisfying to delete so many LOC - thanks for enduring the previous incarnation of this spec!

File: pco/src/tests/recovery.rs [hunk 1]
```diff
@@ -52,21 +52,11 @@ fn assert_recovers<T: NumberLike>(
...
-    for (fm, fq) in [
-      (
-        FloatMultSpec::Enabled,
-        FloatQuantSpec::Disabled,
-      ),
-      (
-        FloatMultSpec::Disabled,
-        FloatQuantSpec::Provided(14),
-      ),
-    ] {
```
Comment (bzinberg) [inline pco/src/tests/recovery.rs:64]: Why the behavior change on this unit test?

File: pco/src/data_types/floats.rs [hunk 2]
```diff
@@ -435,17 +413,16 @@ mod tests {
...
-  // #[test]
-  // fn test_choose_quant_mode() {
-  //   let lowest_num_bits = 1.0_f64.to_bits();
-  //   let k = 20;
-  //   let nums = (0..1000)
-  //     .map(|i| f64::from_bits(lowest_num_bits + (i << k)))
-  //     .collect::<Vec<_>>();
-  //   let (mode, _) = choose_mode_and_split_latents(&nums, &ChunkConfig::default()).unwrap();
-  //   assert_eq!(mode, Mode::FloatQuant(k));
-  // }
```
Comment (bzinberg) [inline pco/src/data_types/floats.rs:448]: How did this test get fixed?  I'm not seeing it, but if the PR contains code that fixes the failing test, can we add that information to the PR description?

**Review by bzinberg — COMMENTED** [review]
Oh yeah this rules.  I'm confused though, see comments

File: pco/src/data_types/floats.rs (hunk 2 above)
Comment (mwlon) [reply pco/src/data_types/floats.rs:448]: I'm being confusing: this was actually fixed in https://github.com/mwlon/pcodec/pull/219, and I should have uncommented the test there, but forgot.

File: pco/src/tests/recovery.rs (hunk 1 above)
Comment (mwlon) [reply pco/src/tests/recovery.rs:64]: At least the first spec here is no longer possible, and I think `Auto + Classic` cover this pretty well. But I might have reduced some coverage with the 2nd one, so maybe I/we should add a new recovery test to hit that case. Thanks for the post-hoc review.

**Pull request merged**